### PR TITLE
Added support for generators as input to GraphAPI.batch

### PR DESCRIPTION
**Pull request by brownstein**
This adds support for generators in GraphAPI.batch.

There's a slight functionality change here as well, which may not be advisable, as the consumer is now responsible for iterating over the output to ensure that all requests for batch sizes over 50 are actually made. Anyway, it seems like functionality that should be somehow supported, although perhaps not be default - let me know.

### CODE REVIEW

File: facepy/graph_api.py [hunk 1]
```diff
@@ -152,51 +152,58 @@ def batch(self, requests):
...
-                yield self._parse(response['body'])
-            except FacepyError as exception:
-                exception.request = request
-                yield exception
+        for request_group in _grouper(requests, 50):
+
+            # url-encode and request bodies
+            for request in request_group:
+                if 'body' in request:
+                    request['body'] = urlencode(request['body'])
```
Comment (jgorset) [inline facepy/graph_api.py]: This change seems unrelated to generators, but more than that; are you sure that we need it?

**Comment by jgorset** [comment]
Is the consumer really required to iterate over the responses for the requests to be made, though? We're iterating through all of the request groups on line 182, which should exhaust the generator.

**Comment by brownstein** [comment]
@jgorset good call - I had written it this way to save memory, but if the consumer fails to iterate through the requests they will indeed not be made. Correcting.

**Comment by brownstein** [comment]
@jgorset fixed PR; all requests should now be made before generator starts returning results.

**Comment by jgorset** [comment]
Nono, look – I'm making the argument that all the requests _would_ be made regardless of whether the consumer iterates through the responses because we're already doing that. Sorry if I wasn't clear.

I'll concede that my Python is a bit rusty, but I'm pretty sure that  5571e6e426636cfc7aaf868b67d7e1c8ea3eec75 doesn't change anything? You're still iterating through the generator until it is exhausted. I don't think we need that intermediary array.

**Comment by brownstein** [comment]
@jgorset simplified changes to logic

File: facepy/graph_api.py [hunk 2]
```diff
@@ -152,33 +152,41 @@ def batch(self, requests):
         """
         Make a batch request.
 
-        :param requests: A list of dictionaries with keys 'method', 'relative_url' and optionally 'body'.
+        :param requests: A list or list comprehension of dictionaries with keys 'method',
```
Comment (merwok) [inline facepy/graph_api.py]: Why not say: An iterable

File: facepy/graph_api.py [hunk 3]
```diff
@@ -152,33 +152,41 @@ def batch(self, requests):
...
         """
 
-        for request in requests:
-            if 'body' in request:
-                request['body'] = urlencode(request['body'])
-
         def _grouper(complete_list, n=1):
             """
-            Batches a list into constant size chunks.
+            Batches a list or generator into constant size chunks.
```
Comment (merwok) [inline facepy/graph_api.py]: generator is a specific type (the type of objects returned by functions that use yield), the generic concept here is iterable.

**Comment by brownstein** [comment]
@merwok Good call, updated 👍

File: facepy/graph_api.py [hunk 4]
```diff
@@ -152,33 +152,41 @@ def batch(self, requests):
         """
         Make a batch request.
 
-        :param requests: A list of dictionaries with keys 'method', 'relative_url' and optionally 'body'.
+        :param requests: An iterable of dictionaries with keys 'method',
+                         'relative_url' and optionally 'body'.
 
-        Yields a list of responses and/or exceptions.
+        Yields a generator of responses and/or exceptions.
```
Comment (merwok) [inline facepy/graph_api.py]: Returns an iterable of*

File: facepy/graph_api.py [hunk 5]
```diff
@@ -152,33 +152,41 @@ def batch(self, requests):
...
-            if 'body' in request:
-                request['body'] = urlencode(request['body'])
-
         def _grouper(complete_list, n=1):
             """
-            Batches a list into constant size chunks.
+            Batches an iterable into constant size chunks.
 
-            :param complete_list: A input list (not a generator).
+            :param complete_list: An input list comprehension.
```
Comment (merwok) [inline facepy/graph_api.py]: This can now be any iterable too

File: facepy/graph_api.py (hunk 4 above)
Comment (brownstein) [inline facepy/graph_api.py]: @merwok fixed!

File: facepy/graph_api.py (hunk 5 above)
Comment (brownstein) [inline facepy/graph_api.py]: @merwok good call!

**Comment by brownstein** [comment]
@merwok docstrings updated to reflect iterable input/output.

**Comment by brownstein** [comment]
I no longer use Facepy, and this has been open for two years. I give up.

**Pull request closed without merging**